// hypertube-nodejs-sdk/lib/plugins/settings/BasePluginSettings.js
/**
 * Represents the base configuration model shared by all plugins.
 * This class contains common metadata such as plugin ID, path, and language type.
 * 
 * @abstract
 * @class BasePluginSettings
 */
export class BasePluginSettings {
  /**
   * Creates an instance of BasePluginSettings.
   * This is a base class and should not be instantiated directly.
   */
  constructor() {
    if (new.target === BasePluginSettings) {
      throw new TypeError('Cannot construct BasePluginSettings instances directly')
    }

    /**
     * A unique identifier assigned by the developer using a predefined plugin.
     * Used to reference or distinguish plugin instances during registration or execution.
     * @type {string}
     */
    this.pluginId = ''

    /**
     * The file system path to the plugin code.
     * This may point to a DLL (for .NET) or other file (for external languages).
     * Should remain empty if the plugin is executed directly from a referenced package.
     * @type {string}
     */
    this.pluginPath = ''

    /**
     * Defines the language or technology in which the plugin is implemented.
     * Used to determine how the plugin should be loaded or invoked.
     * @type {PluginLangType}
     */
    this.pluginLangType = null
  }
}

/**
 * Specifies the programming language or runtime environment used to implement a plugin.
 * @enum {number}
 */
export const PluginLangType = {
  /** .NET Framework (legacy) implementation */
  Clr: 0,

  /** Go implementation */
  Go: 1,

  /** Java class or JAR-based plugin */
  Jvm: 2,

  /** .NET Core / .NET 5+ implementation */
  Netcore: 3,

  /** Perl script or module */
  Perl: 4,

  /** Python script or module */
  Python: 5,

  /** Ruby script or library */
  Ruby: 6,

  /** Node.js JavaScript module */
  Nodejs: 7,

  /** C++ script or class-based plugin */
  Cpp: 8,

  /** PHP script or class-based plugin */
  Php: 9,

  /** Python 2.7 script or class-based plugin */
  Python27: 10
}
