// hypertube-nodejs-sdk/lib/plugins/settings/JwtGraftocodePluginSettings.js
import { BasePluginSettings } from './BasePluginSettings.js'

/**
 * Represents configuration settings specific to JWT plugins,
 * including user identity and authorization logic class information.
 * Inherits shared properties from BasePluginSettings.
 * 
 * @class JwtGraftocodePluginSettings
 * @extends {BasePluginSettings}
 */
export class JwtGraftocodePluginSettings extends BasePluginSettings {
  /**
   * Creates an instance of JwtGraftocodePluginSettings.
   */
  constructor() {
    super()

    /**
     * Username for JWT token generation.
     * @type {string}
     */
    this.username = ''

    /**
     * Secret key used for JWT token signing and validation.
     * @type {string}
     */
    this.secretKey = ''

    /**
     * Whether to use attributes for authorization checks.
     * @type {boolean}
     */
    this.useAttributes = false

    /**
     * JWT authorization class implementation.
     * In Node.js, this is typically a class constructor/function rather than a Type.
     * @type {Function | null}
     */
    this.jwtAuthorizeClass = null
  }
}
