// hypertube-nodejs-sdk/lib/utils/Command.js
// @ts-check
import { CommandType } from './CommandType.js'

/**
 * @typedef {import('../types.d.ts').RuntimeName} RuntimeName
 */

export class Command {
    /**
     * Constructs a new Command instance.
     * @param {RuntimeName} runtimeName - The runtime name associated with the command.
     * @param {number} commandType - The type of the command.
     * @param {any} [payload] - The optional payload of the command.
     * @method
     */
    constructor(runtimeName, commandType, payload = []) {
        this.runtimeName = runtimeName
        this.commandType = commandType
        this.payload = payload
    }

    /**
     * @param {unknown} response
     * @param {RuntimeName} runtimeName
     * @returns {Command}
     */
    static createResponse(response, runtimeName) {
        return new Command(runtimeName, CommandType.Value, [response])
    }

    /**
     * @param {unknown} response
     * @param {RuntimeName} runtimeName
     * @returns {Command}
     */
    static createReference(response, runtimeName) {
        return new Command(runtimeName, CommandType.Reference, [response])
    }

    dropFirstPayloadArg() {
        return new Command(this.runtimeName, this.commandType, this.payload.slice(1))
    }

    /**
     * @param {any} arg
     * @returns {Command}
     */
    addArgToPayload(arg) {
        return new Command(this.runtimeName, this.commandType, this.payload.concat(arg))
    }

    /**
     * @param {Command|null} current_command
     * @returns {Command}
     */
    prependArgToPayload(current_command) {
        if (current_command == null) {
            return new Command(this.runtimeName, this.commandType, this.payload)
        } else {
            return new Command(this.runtimeName, this.commandType, [current_command].concat(this.payload))
        }
    }

    /**
     * Returns string representation of the Command instance for debugging purposes.
     * @returns {string}
     */
    toString() {
        try {
            let result = '  { '
            result += `RuntimeName: ${this.runtimeName}, `
            result += `CommandType: ${this.commandType}, `
            result += 'Payload: \n'
            result += '    ['

            /**
             * @type {string | any[]}
             */
            const payload = this.payload || []
            const len = payload.length

            if (len > 0) {
                result += '\n'
                for (let i = 0; i < len; i++) {
                    const item = payload[i]
                    let itemStr
                    if (item === null || item === undefined) {
                        itemStr = 'null'
                    } else if (typeof item === 'string') {
                        itemStr = `  "${item}"`
                    } else if (item instanceof Command) {
                        itemStr = item.toString().replace(/\n/g, '\n    ')
                    } else {
                        itemStr = '  ' + String(item)
                    }
                    result += '    ' + itemStr
                    result += i < len - 1 ? ',\n' : '\n'
                }
                result += '    ]\n'
            } else {
                result += '  ]\n'
            }

            result += '  }'
            return result
        } catch (ex) {
            return 'Error while converting command to string:' + (ex);
        }
    }
}
