// hypertube-nodejs-sdk/lib/utils/UtilsConst.js
// @ts-check

import { getRequire, isNodejsRuntime } from './Runtime.js'

/** @type {any | null} */
let fs = null
/** @type {any | null} */
let path = null

const requireDynamic = getRequire(import.meta.url)

class UtilsConst {
    /** @type {string} */
    static _workingDirectory = '.'
    /** @type {string} */
    static _configSource = ''
    /** @type {string} */
    static _licenseKey = ''

    /**
     * @param {string} inputPath
     */
    static setWorkingDirectory(inputPath) {
        if (!inputPath) {
            return
        }
        let normalizedPath = inputPath.replace(/\\/g, '/')

        if (!normalizedPath.endsWith('/')) {
            normalizedPath += '/'
        }
        if (isNodejsRuntime()) {
            if (!fs) {
                fs = requireDynamic('fs')
            }
            if (!fs.existsSync(normalizedPath)) {
                fs.mkdirSync(normalizedPath, { recursive: true })
                fs.chmodSync(normalizedPath, 0o700)
            }
        }
        UtilsConst._workingDirectory = normalizedPath
    }

    /**
     * @returns {string}
     */
    static getWorkingDirectory() {
        if (isNodejsRuntime() && !UtilsConst._workingDirectory) {
            if (!path) {
                path = requireDynamic('path')
            }
            UtilsConst._workingDirectory = path.resolve() + '/'
        }
        return UtilsConst._workingDirectory
    }

    /**
     * @param {string} value
     */
    static setConfigSource(value) {
        UtilsConst._configSource = value
    }

    /**
     * @returns {string}
     */
    static getConfigSource() {
        return UtilsConst._configSource
    }

    /**
     * @param {string} value
     */
    static setLicenseKey(value) {
        if (!value || value === 'your-license-key') {
            return
        }
        UtilsConst._licenseKey = value
    }

    /**
     * @returns {string}
     */
    static getLicenseKey() {
        return UtilsConst._licenseKey
    }
}

export { UtilsConst }
