// hypertube-nodejs-sdk/lib/plugins/PluginRegistry.js
/**
 * A thread-safe registry for managing plugins.
 * Provides access to registered plugin instances and their typed variants,
 * allowing for dynamic plugin resolution at runtime.
 * 
 * Note: JavaScript is single-threaded (event loop), so thread-safety considerations
 * from the .NET implementation are not needed here. However, the freeze pattern
 * is still used to prevent modifications after the registration phase.
 */
export class PluginRegistry {
  // Private static fields (ES2022 private fields)
  
  /** @type {Map<string, object>} Map of runtime context IDs to sending plugin instances */
  static #sendingPlugins = new Map()
  
  /** @type {Map<string, object> | null} Immutable snapshot of sending plugins after freeze */
  static #sendingSnapshot = null
  
  /** @type {SendingPluginCacheEntry | null} Hot cache for last accessed plugin */
  static #sendingCache = null
  
  /** @type {object | null} The receiving plugin instance */
  static #receivingPlugin = null
  
  /** @type {Function | null} The receiving plugin constructor/class */
  static #receivingPluginType = null
  
  /** @type {string[] | null} Array of registered runtime context IDs for fast lookup */
  static #registeredKeys = null
  
  /** @type {boolean} Whether to create new plugin instances on each resolve */
  static #useTransientPlugins = false
  
  /** @type {boolean} Whether the registry has been frozen */
  static #isFrozen = false

  /**
   * Checks if a receiving plugin has been registered
   * @returns {boolean} True if a receiving plugin exists
   */
  static get hasReceivingPlugin() {
    return PluginRegistry.#receivingPlugin !== null
  }

  /**
   * Registers a sending plugin implementation for a specific runtime context.
   * The plugin is stored as a singleton instance and can later be resolved using the runtime context ID.
   * 
   * @param {string} runtimeContextId - The UUID of the runtime context
   * @param {object} plugin - Plugin instance that implements ICommunitySendingPlugin interface
   * @throws {Error} If registry is already frozen
   * @throws {TypeError} If plugin doesn't have required execute method
   * 
   * @example
   * const plugin = new JwtPlugin('token123')
   * PluginRegistry.registerSendingPlugin(runtimeCtx.runtimeContextId, plugin)
   */
  static registerSendingPlugin(runtimeContextId, plugin) {
    if (PluginRegistry.#isFrozen) {
      throw new Error('Cannot register plugins after registry has been frozen. Call reset() first.')
    }

    if (!plugin || typeof plugin.execute !== 'function') {
      throw new TypeError('Plugin must have an execute() method')
    }

    if (!plugin.pluginId) {
      throw new TypeError('Plugin must have a pluginId property')
    }

    if (!runtimeContextId || typeof runtimeContextId !== 'string') {
      throw new TypeError('runtimeContextId must be a non-empty string')
    }

    PluginRegistry.#sendingPlugins.set(runtimeContextId, plugin)

    // Update cache with the newly registered plugin
    PluginRegistry.#sendingCache = {
      key: runtimeContextId,
      value: plugin,
      hasValue: true,
      type: plugin.constructor
    }
  }

  /**
   * Registers a receiving plugin implementation.
   * The plugin is stored as a singleton instance and executed on the receiving side.
   * Only one receiving plugin can be registered per application.
   * 
   * @param {object} plugin - Plugin instance that implements ICommunityReceivingPlugin interface
   * @throws {Error} If a receiving plugin is already registered
   * @throws {Error} If registry is already frozen
   * @throws {TypeError} If plugin doesn't have required execute method
   * 
   * @example
   * const plugin = new ValidationPlugin()
   * PluginRegistry.registerReceivingPlugin(plugin)
   */
  static registerReceivingPlugin(plugin) {
    if (PluginRegistry.#isFrozen) {
      throw new Error('Cannot register plugins after registry has been frozen. Call reset() first.')
    }

    if (PluginRegistry.#receivingPlugin !== null) {
      throw new Error('A receiving plugin is already registered. Only one receiving plugin is allowed.')
    }

    if (!plugin || typeof plugin.execute !== 'function') {
      throw new TypeError('Plugin must have an execute(commandPayload, executionTarget) method')
    }

    if (!plugin.pluginId) {
      throw new TypeError('Plugin must have a pluginId property')
    }

    PluginRegistry.#receivingPlugin = plugin
    PluginRegistry.#receivingPluginType = plugin.constructor
  }

  /**
   * Resolves and returns the registered receiving plugin instance.
   * If transient mode is enabled, creates a new instance.
   * 
   * @returns {object} The receiving plugin instance
   * @throws {Error} If no receiving plugin has been registered
   * 
   * @example
   * const plugin = PluginRegistry.resolveReceivingPlugin()
   * const result = plugin.execute(payload, target)
   */
  static resolveReceivingPlugin() {
    if (PluginRegistry.#useTransientPlugins && PluginRegistry.#receivingPluginType) {
      return new PluginRegistry.#receivingPluginType()
    }

    if (PluginRegistry.#receivingPlugin === null) {
      throw new Error('No receiving plugin has been registered')
    }

    return PluginRegistry.#receivingPlugin
  }

  /**
   * Attempts to retrieve a sending plugin registered for the specified runtime context.
   * First checks a single-entry hot cache for the last accessed plugin,
   * and if not found, falls back to the map of registered plugins.
   * On a successful map lookup, the cache is updated with the resolved entry
   * for faster access in subsequent calls.
   * 
   * @param {string} runtimeContextId - The identifier of the runtime context
   * @returns {{success: boolean, plugin: object | null}} Object with success flag and plugin instance
   * 
   * @example
   * const result = PluginRegistry.tryGetSendingPlugin(runtimeContextId)
   * if (result.success) {
   *   const payload = result.plugin.execute()
   * }
   */
  static tryGetSendingPlugin(runtimeContextId) {
    const localCache = PluginRegistry.#sendingCache

    // If using transient plugins, create a new instance
    if (PluginRegistry.#useTransientPlugins && localCache && localCache.hasValue && localCache.key === runtimeContextId) {
      return {
        success: true,
        plugin: new localCache.type()
      }
    }

    // Check hot cache first
    if (localCache && localCache.hasValue && localCache.key === runtimeContextId) {
      return {
        success: true,
        plugin: localCache.value
      }
    }

    // Fall back to snapshot/map
    const snapshot = PluginRegistry.#sendingSnapshot
    if (snapshot && snapshot.has(runtimeContextId)) {
      const plugin = snapshot.get(runtimeContextId)
      
      // Update cache for next access
      PluginRegistry.#sendingCache = {
        key: runtimeContextId,
        value: plugin,
        hasValue: true,
        type: plugin.constructor
      }

      return {
        success: true,
        plugin: plugin
      }
    }

    return {
      success: false,
      plugin: null
    }
  }

  /**
   * Checks whether any plugins have been registered for given RuntimeContext ID.
   * 
   * @param {string} runtimeContextId - The identifier of the runtime context
   * @returns {boolean} True if at least one plugin is registered for the runtime context
   * 
   * @example
   * if (PluginRegistry.arePluginsRegisteredForRuntime(runtimeContextId)) {
   *   // Build and wrap command with plugin payload
   * }
   */
  static arePluginsRegisteredForRuntime(runtimeContextId) {
    const keys = PluginRegistry.#registeredKeys

    if (!keys || keys.length === 0) {
      return false
    }

    // Fast array lookup
    return keys.includes(runtimeContextId)
  }

  /**
   * Finalizes the plugin registration phase by creating a compact snapshot of all registered keys.
   * This snapshot is stored to allow extremely fast lookups on the hot path.
   * After calling freeze(), no more plugins can be registered unless reset() is called.
   * 
   * @param {boolean} useTransientPlugins - If true, creates new plugin instances on each resolve
   * 
   * @example
   * // Register all plugins
   * PluginRegistry.registerSendingPlugin(rtmCtx.runtimeContextId, plugin1)
   * PluginRegistry.registerSendingPlugin(rtmCtx2.runtimeContextId, plugin2)
   * 
   * // Freeze the registry
   * PluginRegistry.freeze()
   * 
   * // Or freeze with transient mode
   * PluginRegistry.freeze(true)
   */
  static freeze(useTransientPlugins = false) {
    PluginRegistry.#useTransientPlugins = useTransientPlugins

    // Create array of registered keys for fast lookup
    PluginRegistry.#registeredKeys = Array.from(PluginRegistry.#sendingPlugins.keys())

    // Create immutable snapshot
    PluginRegistry.#sendingSnapshot = new Map(PluginRegistry.#sendingPlugins)

    // Clear the mutable map to prevent accidental modifications
    PluginRegistry.#sendingPlugins.clear()

    // Mark as frozen
    PluginRegistry.#isFrozen = true
  }

  /**
   * Resets the entire plugin registry to its initial state.
   * Clears all registered plugins and unfreezes the registry.
   * Useful for testing or reconfiguration scenarios.
   * 
   * @example
   * PluginRegistry.reset()
   * // Can now register new plugins
   */
  static reset() {
    PluginRegistry.#sendingPlugins.clear()
    PluginRegistry.#sendingSnapshot = null
    PluginRegistry.#sendingCache = null
    PluginRegistry.#receivingPlugin = null
    PluginRegistry.#receivingPluginType = null
    PluginRegistry.#registeredKeys = null
    PluginRegistry.#useTransientPlugins = false
    PluginRegistry.#isFrozen = false
  }

  /**
   * Checks if the registry has been frozen
   * @returns {boolean} True if the registry is frozen
   */
  static get isFrozen() {
    return PluginRegistry.#isFrozen
  }
}

/**
 * @typedef {Object} SendingPluginCacheEntry
 * @property {string} key - The runtime context ID
 * @property {object} value - The plugin instance
 * @property {boolean} hasValue - Whether the entry has a value
 * @property {Function} type - The plugin constructor/class
 */
