// hypertube-nodejs-sdk/lib/plugins/PluginImplementationRegistry.js
/**
 * A registry for managing plugins and their implementations.
 * Provides access to registered plugin instances and their typed variants,
 * allowing for dynamic plugin resolution at runtime.
 * 
 * This class acts as a lightweight alternative to a dependency injection container,
 * supporting both singleton and transient lifecycles for plugins.
 * Plugins can be resolved by type, and optionally created with constructor arguments.
 */
export class PluginImplementationRegistry {
  /** @type {Map<string, Function>} Map of interface names to implementation constructors for transient plugins */
  static #transientImplementations = new Map()
  
  /** @type {Map<Function, object>} Map of implementation types to singleton instances */
  static #singletonImplementations = new Map()
  
  /** @type {Map<Function, object>} Map of settings types to configuration objects */
  static #pluginConfigurations = new Map()

  /**
   * Registers a transient plugin implementation. A new instance will be created each time the plugin is resolved.
   * 
   * @param {Function} InterfaceClass - The interface or base class of the plugin (must be a class/constructor)
   * @param {Function} ImplementationClass - The concrete class that implements InterfaceClass
   * @throws {TypeError} If InterfaceClass is not a function/constructor
   * @throws {TypeError} If ImplementationClass is not a function/constructor
   * 
   * @example
   * PluginImplementationRegistry.registerPluginTransientImplementation(IJwtAuthorize, JwtAuthorizeImpl)
   */
  static registerPluginTransientImplementation(InterfaceClass, ImplementationClass) {
    if (typeof InterfaceClass !== 'function') {
      throw new TypeError(`${InterfaceClass?.name || 'InterfaceClass'} is not a function. Only classes/constructors are allowed.`)
    }

    if (typeof ImplementationClass !== 'function') {
      throw new TypeError(`${ImplementationClass?.name || 'ImplementationClass'} is not a function. Only classes/constructors are allowed.`)
    }

    // Use the interface class name as the key
    const interfaceName = InterfaceClass.name || InterfaceClass.toString()
    PluginImplementationRegistry.#transientImplementations.set(interfaceName, ImplementationClass)
  }

  /**
   * Resolves a new instance of a transiently registered plugin.
   * 
   * @param {Function} InterfaceClass - The type of the plugin interface to resolve
   * @param {...any} constructorArgs - Optional arguments to pass to the constructor
   * @returns {object} A newly created instance of the registered implementation
   * @throws {Error} If the plugin type has not been registered as transient
   * @throws {TypeError} If InterfaceClass is not a function/constructor
   * 
   * @example
   * const jwtAuthorize = PluginImplementationRegistry.resolvePluginTransientImplementation(IJwtAuthorize)
   */
  static resolvePluginTransientImplementation(InterfaceClass, ...constructorArgs) {
    if (typeof InterfaceClass !== 'function') {
      throw new TypeError(`${InterfaceClass?.name || 'InterfaceClass'} is not a function. Only classes/constructors are allowed.`)
    }

    const interfaceName = InterfaceClass.name || InterfaceClass.toString()
    const ImplementationClass = PluginImplementationRegistry.#transientImplementations.get(interfaceName)

    if (!ImplementationClass) {
      throw new Error(`No transient implementation registered for ${interfaceName}.`)
    }

    try {
      // Create new instance with constructor arguments if provided
      const instance = constructorArgs.length > 0
        ? new ImplementationClass(...constructorArgs)
        : new ImplementationClass()

      if (!instance) {
        throw new Error(`Could not instantiate ${ImplementationClass.name || 'implementation'}.`)
      }

      return instance
    } catch (error) {
      throw new Error(`Failed to create instance of ${ImplementationClass.name || 'implementation'}: ${error.message}`)
    }
  }

  /**
   * Registers a plugin configuration/settings object.
   * 
   * @param {Function} SettingsClass - The settings class type
   * @param {object} settings - The settings instance to register
   * @throws {TypeError} If SettingsClass is not a function/constructor
   * @throws {TypeError} If settings is not an object
   * 
   * @example
   * const settings = new JwtGraftocodePluginSettings()
   * settings.username = 'user'
   * settings.secretKey = 'key'
   * PluginImplementationRegistry.registerPluginConfiguration(JwtGraftocodePluginSettings, settings)
   */
  static registerPluginConfiguration(SettingsClass, settings) {
    if (typeof SettingsClass !== 'function') {
      throw new TypeError(`${SettingsClass?.name || 'SettingsClass'} is not a function. Only classes/constructors are allowed.`)
    }

    if (!settings || typeof settings !== 'object') {
      throw new TypeError('settings must be an object')
    }

    PluginImplementationRegistry.#pluginConfigurations.set(SettingsClass, settings)
  }

  /**
   * Retrieves the configuration object of the specified type for a given plugin.
   * 
   * @param {Function} SettingsClass - The expected configuration type
   * @returns {object | null} The configuration object, or null if not found
   * @throws {TypeError} If SettingsClass is not a function/constructor
   * 
   * @example
   * const config = PluginImplementationRegistry.getPluginConfiguration(JwtGraftocodePluginSettings)
   * if (config) {
   *   console.log(config.username)
   * }
   */
  static getPluginConfiguration(SettingsClass) {
    if (typeof SettingsClass !== 'function') {
      throw new TypeError(`${SettingsClass?.name || 'SettingsClass'} is not a function. Only classes/constructors are allowed.`)
    }

    const config = PluginImplementationRegistry.#pluginConfigurations.get(SettingsClass)
    return config || null
  }

  /**
   * Checks whether a configuration has been registered for the specified plugin type.
   * 
   * @param {Function} SettingsClass - The settings class type to check
   * @returns {boolean} True if a configuration entry exists, false otherwise
   * @throws {TypeError} If SettingsClass is not a function/constructor
   * 
   * @example
   * if (PluginImplementationRegistry.isPluginConfigurationAvailable(JwtGraftocodePluginSettings)) {
   *   // Configuration exists
   * }
   */
  static isPluginConfigurationAvailable(SettingsClass) {
    if (typeof SettingsClass !== 'function') {
      throw new TypeError(`${SettingsClass?.name || 'SettingsClass'} is not a function. Only classes/constructors are allowed.`)
    }

    return PluginImplementationRegistry.#pluginConfigurations.has(SettingsClass)
  }

  /**
   * Registers a new plugin implementation of the specified type and stores it as a singleton instance.
   * 
   * @param {Function} InterfaceClass - The interface or base class that the plugin implementation inherits or implements
   * @param {Function} ImplementationClass - The concrete plugin type to register
   * @param {...any} constructorArgs - Optional arguments to pass to the constructor
   * @throws {TypeError} If InterfaceClass is not a function/constructor
   * @throws {TypeError} If ImplementationClass is not a function/constructor
   * 
   * @example
   * PluginImplementationRegistry.registerPluginSingletonImplementation(IJwtAuthorize, JwtAuthorizeImpl)
   * // Or with constructor arguments:
   * PluginImplementationRegistry.registerPluginSingletonImplementation(IJwtAuthorize, JwtAuthorizeImpl, 'arg1', 'arg2')
   */
  static registerPluginSingletonImplementation(InterfaceClass, ImplementationClass, ...constructorArgs) {
    if (typeof InterfaceClass !== 'function') {
      throw new TypeError(`${InterfaceClass?.name || 'InterfaceClass'} is not a function. Only classes/constructors are allowed.`)
    }

    if (typeof ImplementationClass !== 'function') {
      throw new TypeError(`${ImplementationClass?.name || 'ImplementationClass'} is not a function. Only classes/constructors are allowed.`)
    }

    try {
      // Create instance with constructor arguments if provided
      const instance = constructorArgs.length > 0
        ? new ImplementationClass(...constructorArgs)
        : new ImplementationClass()

      if (!instance) {
        throw new Error(`Could not create instance of ${ImplementationClass.name || 'implementation'}.`)
      }

      PluginImplementationRegistry.#singletonImplementations.set(ImplementationClass, instance)
    } catch (error) {
      throw new Error(`Failed to register singleton implementation: ${error.message}`)
    }
  }

  /**
   * Resolves a registered plugin implementation by its type.
   * 
   * @param {Function} ImplementationClass - The type of the plugin implementation to resolve
   * @returns {object} The registered plugin instance
   * @throws {Error} If no plugin instance of the specified type is found in the registry
   * @throws {TypeError} If ImplementationClass is not a function/constructor
   * 
   * @example
   * const plugin = PluginImplementationRegistry.resolvePluginSingletonImplementation(JwtAuthorizeImpl)
   */
  static resolvePluginSingletonImplementation(ImplementationClass) {
    if (typeof ImplementationClass !== 'function') {
      throw new TypeError(`${ImplementationClass?.name || 'ImplementationClass'} is not a function. Only classes/constructors are allowed.`)
    }

    const instance = PluginImplementationRegistry.#singletonImplementations.get(ImplementationClass)

    if (!instance) {
      throw new Error(`No singleton implementation registered for ${ImplementationClass.name || 'implementation'}.`)
    }

    return instance
  }

  /**
   * Resets the entire plugin implementation registry to its initial state.
   * Clears all registered plugins and configurations.
   * Useful for testing or reconfiguration scenarios.
   * 
   * @example
   * PluginImplementationRegistry.reset()
   * // Can now register new plugins
   */
  static reset() {
    PluginImplementationRegistry.#transientImplementations.clear()
    PluginImplementationRegistry.#singletonImplementations.clear()
    PluginImplementationRegistry.#pluginConfigurations.clear()
  }
}
