// hypertube-nodejs-sdk/dist/plugins/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var plugins_exports = {};
__export(plugins_exports, {
  BasePluginSettings: () => import_BasePluginSettings.BasePluginSettings,
  ICommunityPlugin: () => import_ICommunityPlugin.ICommunityPlugin,
  ICommunityReceivingPlugin: () => import_ICommunityReceivingPlugin.ICommunityReceivingPlugin,
  ICommunitySendingPlugin: () => import_ICommunitySendingPlugin.ICommunitySendingPlugin,
  JwtGraftocodePluginSettings: () => import_JwtGraftocodePluginSettings.JwtGraftocodePluginSettings,
  PluginImplementationRegistry: () => import_PluginImplementationRegistry.PluginImplementationRegistry,
  PluginLangType: () => import_BasePluginSettings.PluginLangType,
  PluginPayloadBuilder: () => import_PluginPayloadBuilder.PluginPayloadBuilder,
  PluginRegistry: () => import_PluginRegistry.PluginRegistry
});
module.exports = __toCommonJS(plugins_exports);
var import_PluginRegistry = require("./PluginRegistry.cjs");
var import_PluginPayloadBuilder = require("./PluginPayloadBuilder.cjs");
var import_ICommunityPlugin = require("./interfaces/ICommunityPlugin.cjs");
var import_ICommunitySendingPlugin = require("./interfaces/ICommunitySendingPlugin.cjs");
var import_ICommunityReceivingPlugin = require("./interfaces/ICommunityReceivingPlugin.cjs");
var import_PluginImplementationRegistry = require("./PluginImplementationRegistry.cjs");
var import_BasePluginSettings = require("./settings/BasePluginSettings.cjs");
var import_JwtGraftocodePluginSettings = require("./settings/JwtGraftocodePluginSettings.cjs");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BasePluginSettings,
  ICommunityPlugin,
  ICommunityReceivingPlugin,
  ICommunitySendingPlugin,
  JwtGraftocodePluginSettings,
  PluginImplementationRegistry,
  PluginLangType,
  PluginPayloadBuilder,
  PluginRegistry
});
