// hypertube-nodejs-sdk/dist/plugins/PluginRegistry.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var PluginRegistry_exports = {};
__export(PluginRegistry_exports, {
  PluginRegistry: () => PluginRegistry
});
module.exports = __toCommonJS(PluginRegistry_exports);
class PluginRegistry {
  // Private static fields (ES2022 private fields)
  /** @type {Map<string, object>} Map of runtime context IDs to sending plugin instances */
  static #sendingPlugins = /* @__PURE__ */ new Map();
  /** @type {Map<string, object> | null} Immutable snapshot of sending plugins after freeze */
  static #sendingSnapshot = null;
  /** @type {SendingPluginCacheEntry | null} Hot cache for last accessed plugin */
  static #sendingCache = null;
  /** @type {object | null} The receiving plugin instance */
  static #receivingPlugin = null;
  /** @type {Function | null} The receiving plugin constructor/class */
  static #receivingPluginType = null;
  /** @type {string[] | null} Array of registered runtime context IDs for fast lookup */
  static #registeredKeys = null;
  /** @type {boolean} Whether to create new plugin instances on each resolve */
  static #useTransientPlugins = false;
  /** @type {boolean} Whether the registry has been frozen */
  static #isFrozen = false;
  /**
   * Checks if a receiving plugin has been registered
   * @returns {boolean} True if a receiving plugin exists
   */
  static get hasReceivingPlugin() {
    return PluginRegistry.#receivingPlugin !== null;
  }
  /**
   * Registers a sending plugin implementation for a specific runtime context.
   * The plugin is stored as a singleton instance and can later be resolved using the runtime context ID.
   * 
   * @param {string} runtimeContextId - The UUID of the runtime context
   * @param {object} plugin - Plugin instance that implements ICommunitySendingPlugin interface
   * @throws {Error} If registry is already frozen
   * @throws {TypeError} If plugin doesn't have required execute method
   * 
   * @example
   * const plugin = new JwtPlugin('token123')
   * PluginRegistry.registerSendingPlugin(runtimeCtx.runtimeContextId, plugin)
   */
  static registerSendingPlugin(runtimeContextId, plugin) {
    if (PluginRegistry.#isFrozen) {
      throw new Error("Cannot register plugins after registry has been frozen. Call reset() first.");
    }
    if (!plugin || typeof plugin.execute !== "function") {
      throw new TypeError("Plugin must have an execute() method");
    }
    if (!plugin.pluginId) {
      throw new TypeError("Plugin must have a pluginId property");
    }
    if (!runtimeContextId || typeof runtimeContextId !== "string") {
      throw new TypeError("runtimeContextId must be a non-empty string");
    }
    PluginRegistry.#sendingPlugins.set(runtimeContextId, plugin);
    PluginRegistry.#sendingCache = {
      key: runtimeContextId,
      value: plugin,
      hasValue: true,
      type: plugin.constructor
    };
  }
  /**
   * Registers a receiving plugin implementation.
   * The plugin is stored as a singleton instance and executed on the receiving side.
   * Only one receiving plugin can be registered per application.
   * 
   * @param {object} plugin - Plugin instance that implements ICommunityReceivingPlugin interface
   * @throws {Error} If a receiving plugin is already registered
   * @throws {Error} If registry is already frozen
   * @throws {TypeError} If plugin doesn't have required execute method
   * 
   * @example
   * const plugin = new ValidationPlugin()
   * PluginRegistry.registerReceivingPlugin(plugin)
   */
  static registerReceivingPlugin(plugin) {
    if (PluginRegistry.#isFrozen) {
      throw new Error("Cannot register plugins after registry has been frozen. Call reset() first.");
    }
    if (PluginRegistry.#receivingPlugin !== null) {
      throw new Error("A receiving plugin is already registered. Only one receiving plugin is allowed.");
    }
    if (!plugin || typeof plugin.execute !== "function") {
      throw new TypeError("Plugin must have an execute(commandPayload, executionTarget) method");
    }
    if (!plugin.pluginId) {
      throw new TypeError("Plugin must have a pluginId property");
    }
    PluginRegistry.#receivingPlugin = plugin;
    PluginRegistry.#receivingPluginType = plugin.constructor;
  }
  /**
   * Resolves and returns the registered receiving plugin instance.
   * If transient mode is enabled, creates a new instance.
   * 
   * @returns {object} The receiving plugin instance
   * @throws {Error} If no receiving plugin has been registered
   * 
   * @example
   * const plugin = PluginRegistry.resolveReceivingPlugin()
   * const result = plugin.execute(payload, target)
   */
  static resolveReceivingPlugin() {
    if (PluginRegistry.#useTransientPlugins && PluginRegistry.#receivingPluginType) {
      return new PluginRegistry.#receivingPluginType();
    }
    if (PluginRegistry.#receivingPlugin === null) {
      throw new Error("No receiving plugin has been registered");
    }
    return PluginRegistry.#receivingPlugin;
  }
  /**
   * Attempts to retrieve a sending plugin registered for the specified runtime context.
   * First checks a single-entry hot cache for the last accessed plugin,
   * and if not found, falls back to the map of registered plugins.
   * On a successful map lookup, the cache is updated with the resolved entry
   * for faster access in subsequent calls.
   * 
   * @param {string} runtimeContextId - The identifier of the runtime context
   * @returns {{success: boolean, plugin: object | null}} Object with success flag and plugin instance
   * 
   * @example
   * const result = PluginRegistry.tryGetSendingPlugin(runtimeContextId)
   * if (result.success) {
   *   const payload = result.plugin.execute()
   * }
   */
  static tryGetSendingPlugin(runtimeContextId) {
    const localCache = PluginRegistry.#sendingCache;
    if (PluginRegistry.#useTransientPlugins && localCache && localCache.hasValue && localCache.key === runtimeContextId) {
      return {
        success: true,
        plugin: new localCache.type()
      };
    }
    if (localCache && localCache.hasValue && localCache.key === runtimeContextId) {
      return {
        success: true,
        plugin: localCache.value
      };
    }
    const snapshot = PluginRegistry.#sendingSnapshot;
    if (snapshot && snapshot.has(runtimeContextId)) {
      const plugin = snapshot.get(runtimeContextId);
      PluginRegistry.#sendingCache = {
        key: runtimeContextId,
        value: plugin,
        hasValue: true,
        type: plugin.constructor
      };
      return {
        success: true,
        plugin
      };
    }
    return {
      success: false,
      plugin: null
    };
  }
  /**
   * Checks whether any plugins have been registered for given RuntimeContext ID.
   * 
   * @param {string} runtimeContextId - The identifier of the runtime context
   * @returns {boolean} True if at least one plugin is registered for the runtime context
   * 
   * @example
   * if (PluginRegistry.arePluginsRegisteredForRuntime(runtimeContextId)) {
   *   // Build and wrap command with plugin payload
   * }
   */
  static arePluginsRegisteredForRuntime(runtimeContextId) {
    const keys = PluginRegistry.#registeredKeys;
    if (!keys || keys.length === 0) {
      return false;
    }
    return keys.includes(runtimeContextId);
  }
  /**
   * Finalizes the plugin registration phase by creating a compact snapshot of all registered keys.
   * This snapshot is stored to allow extremely fast lookups on the hot path.
   * After calling freeze(), no more plugins can be registered unless reset() is called.
   * 
   * @param {boolean} useTransientPlugins - If true, creates new plugin instances on each resolve
   * 
   * @example
   * // Register all plugins
   * PluginRegistry.registerSendingPlugin(rtmCtx.runtimeContextId, plugin1)
   * PluginRegistry.registerSendingPlugin(rtmCtx2.runtimeContextId, plugin2)
   * 
   * // Freeze the registry
   * PluginRegistry.freeze()
   * 
   * // Or freeze with transient mode
   * PluginRegistry.freeze(true)
   */
  static freeze(useTransientPlugins = false) {
    PluginRegistry.#useTransientPlugins = useTransientPlugins;
    PluginRegistry.#registeredKeys = Array.from(PluginRegistry.#sendingPlugins.keys());
    PluginRegistry.#sendingSnapshot = new Map(PluginRegistry.#sendingPlugins);
    PluginRegistry.#sendingPlugins.clear();
    PluginRegistry.#isFrozen = true;
  }
  /**
   * Resets the entire plugin registry to its initial state.
   * Clears all registered plugins and unfreezes the registry.
   * Useful for testing or reconfiguration scenarios.
   * 
   * @example
   * PluginRegistry.reset()
   * // Can now register new plugins
   */
  static reset() {
    PluginRegistry.#sendingPlugins.clear();
    PluginRegistry.#sendingSnapshot = null;
    PluginRegistry.#sendingCache = null;
    PluginRegistry.#receivingPlugin = null;
    PluginRegistry.#receivingPluginType = null;
    PluginRegistry.#registeredKeys = null;
    PluginRegistry.#useTransientPlugins = false;
    PluginRegistry.#isFrozen = false;
  }
  /**
   * Checks if the registry has been frozen
   * @returns {boolean} True if the registry is frozen
   */
  static get isFrozen() {
    return PluginRegistry.#isFrozen;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PluginRegistry
});
