// hypertube-nodejs-sdk/dist/plugins/settings/BasePluginSettings.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var BasePluginSettings_exports = {};
__export(BasePluginSettings_exports, {
  BasePluginSettings: () => BasePluginSettings,
  PluginLangType: () => PluginLangType
});
module.exports = __toCommonJS(BasePluginSettings_exports);
class BasePluginSettings {
  /**
   * Creates an instance of BasePluginSettings.
   * This is a base class and should not be instantiated directly.
   */
  constructor() {
    if (new.target === BasePluginSettings) {
      throw new TypeError("Cannot construct BasePluginSettings instances directly");
    }
    this.pluginId = "";
    this.pluginPath = "";
    this.pluginLangType = null;
  }
}
const PluginLangType = {
  /** .NET Framework (legacy) implementation */
  Clr: 0,
  /** Go implementation */
  Go: 1,
  /** Java class or JAR-based plugin */
  Jvm: 2,
  /** .NET Core / .NET 5+ implementation */
  Netcore: 3,
  /** Perl script or module */
  Perl: 4,
  /** Python script or module */
  Python: 5,
  /** Ruby script or library */
  Ruby: 6,
  /** Node.js JavaScript module */
  Nodejs: 7,
  /** C++ script or class-based plugin */
  Cpp: 8,
  /** PHP script or class-based plugin */
  Php: 9,
  /** Python 2.7 script or class-based plugin */
  Python27: 10
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  BasePluginSettings,
  PluginLangType
});
