// hypertube-nodejs-sdk/lib/sdk/configuration/Config.js
// @ts-check

class Config {
    /**
     * @param {*} runtime
     * @param {*} connectionData
     * @param {boolean} [isStateless=false]
     * @param {string} [plugins=""]
     * @param {string} [modules=""]
     */
    constructor(runtime, connectionData, isStateless = false, plugins = '', modules = '') {
        this.runtime = runtime
        this.connectionData = connectionData
        this.isStateless = isStateless === true
        this.plugins = plugins ?? ''
        this.modules = modules ?? ''
    }

    toString() {
        const parts = []
        parts.push(`Runtime: ${this.runtime}`)
        parts.push(`IsStateless: ${this.isStateless}`)

        if (this.connectionData != null) {
            parts.push(`Host: ${this.connectionData}`)
        }

        if (String(this.plugins).trim()) {
            parts.push(`Plugins: ${this.plugins}`)
        }

        if (String(this.modules).trim()) {
            parts.push(`Modules: ${this.modules}`)
        }

        return parts.join(', ')
    }
}

export { Config }
